### PR TITLE
Added OpenAppID feature.

### PR DESCRIPTION
**Pull request by rbshadow**
Integrating **OpenAppID** ( Application Detector Package )

**Log directory:** _`/var/log/snort/`_
**Run Command:** `sudo u2openappid /var/log/snort/appstats-u2.log.1393807981` ( _Your log id should not be the same_ )

Output: { _Sample_ }

`statTime="1393807860",appName="chrome",txBytes="6043",rxBytes="111267"
statTime="1393807860",appName="dns",txBytes="8708",rxBytes="38103"
statTime="1393807860",appName="http",txBytes="200399",rxBytes="1444070"
statTime="1393807860",appName="cnn.com",txBytes="198478",rxBytes="1557970"
statTime="1393807860",appName="doubleclick",txBytes="5543",rxBytes="2598"
statTime="1393807860",appName="truste",txBytes="1829",rxBytes="12208"
statTime="1393807860",appName="washington_time",txBytes="2210",rxBytes="1401"`

### CODE REVIEW

**Comment by rbshadow** [comment]
Thank you very much.... I will make the change insha Allah. Give me time..

File: src/Snorter_Ubuntu-14.04.sh [hunk 1]
```diff
@@ -97,21 +134,22 @@ function snort_edit() {
 	echo -ne "\n\t${YELLOW}[!] WARNING:${NOCOLOR} Press ${BOLD}ENTER${NOCOLOR} to continue. "
 	read -n 1 -s
 	sudo vim /etc/snort/snort.conf -c "/ipvar EXTERNAL_NET"
-
+	echo -ne "\n\t${YELLOW}[!] WARNING:${NOCOLOR} Press ${BOLD}ENTER${NOCOLOR} to continue. "
```
Comment (rbshadow) [inline src/Snorter_Ubuntu-14.04.sh]: Mistakenly it happens... :(

**Comment by joanbono** [comment]
Hei @rbshadow , are you working on this?

**Comment by rbshadow** [comment]
Hey @joanbono bro actually I was in EiD vacation :). Today I am back in my workstation. Hope within some days I can complete.

File: src/Snorter_Ubuntu-14.04.sh [hunk 2]
```diff
@@ -22,13 +24,31 @@ function update_upgrade() {
 	
 }
 
+function nghttp_insatll() {
```
Comment (joanbono) [inline src/Snorter_Ubuntu-14.04.sh]: + [x] Done

File: src/Snorter_Ubuntu-14.04.sh [hunk 3]
```diff
@@ -22,13 +24,31 @@ function update_upgrade() {
...
+
+	echo -ne "\n\t${CYAN}[i] INFO:${NOCOLOR} Installing development libraries. \n\n"
+	sudo apt-get install -y --force-yes autoconf libtool pkg-config
```
Comment (joanbono) [inline src/Snorter_Ubuntu-14.04.sh]: + [x] Done

File: src/Snorter_Ubuntu-14.04.sh [hunk 4]
```diff
@@ -22,13 +24,31 @@ function update_upgrade() {
...
+	cd $HOME && mkdir snort_src && cd snort_src
```
Comment (joanbono) [inline src/Snorter_Ubuntu-14.04.sh]: + [x] Done

File: src/Snorter_Ubuntu-14.04.sh [hunk 5]
```diff
@@ -22,13 +24,31 @@ function update_upgrade() {
...
+	echo -ne "\n\t${CYAN}[i] INFO:${NOCOLOR} Downloading Nghttp${BOLD}$NGHTTP2${NOCOLOR}.\n\n"
+	wget --no-check-certificate -P $HOME/snort_src https://github.com/nghttp2/nghttp2/releases/download/v1.23.1/nghttp2-1.23.1.tar.gz
```
Comment (joanbono) [inline src/Snorter_Ubuntu-14.04.sh]: + [x] Done

File: src/Snorter_Ubuntu-14.04.sh [hunk 6]
```diff
@@ -22,13 +24,31 @@ function update_upgrade() {
...
+	cd $HOME/snort_src/
+	tar -xzvf nghttp2-1.23.1.tar.gz
```
Comment (joanbono) [inline src/Snorter_Ubuntu-14.04.sh]: + [x] Done

File: src/Snorter_Ubuntu-14.04.sh [hunk 7]
```diff
@@ -22,13 +24,31 @@ function update_upgrade() {
...
+	./configure --enable-lib-only
+	make
+	sudo make install
+}
+
 function snort_install() {
 
 	echo -ne "\n\t${CYAN}[i] INFO:${NOCOLOR} Installing dependencies.\n\n"
-	sudo apt-get install -y --force-yes build-essential libpcap-dev libpcre3-dev libdumbnet-dev bison flex zlib1g-dev git locate vim
+	sudo apt-get install -y --force-yes build-essential libpcap-dev libpcre3-dev libdumbnet-dev bison flex zlib1g-dev git locate vim libluajit-5.1-dev pkg-config openssl libssl-dev
```
Comment (joanbono) [inline src/Snorter_Ubuntu-14.04.sh]: + [x] Done

File: src/Snorter_Ubuntu-14.04.sh [hunk 8]
```diff
@@ -22,13 +24,31 @@ function update_upgrade() {
...
 	
 	#Downloading DAQ and SNORT
-	cd $HOME && mkdir snort_src && cd snort_src
```
Comment (joanbono) [inline src/Snorter_Ubuntu-14.04.sh]: + [x] Done

File: src/Snorter_Ubuntu-14.04.sh [hunk 9]
```diff
@@ -44,13 +64,13 @@ function snort_install() {
...
 	#Installing SNORT
-	cd $HOME/snort_src
+	cd $HOME/snort_src/
 	echo -ne "\n\t${CYAN}[i] INFO:${NOCOLOR} Installing ${BOLD}$SNORT${NOCOLOR}.\n\n"
 	tar xvfz $SNORT.tar.gz > /dev/null 2>&1
 	rm -r *.tar.gz > /dev/null 2>&1
 	mv snort-*/ snort           
 	cd snort
-	./configure --enable-sourcefire && make && sudo make install
+	./configure --enable-sourcefire --enable-open-appid && make && sudo make install
```
Comment (joanbono) [inline src/Snorter_Ubuntu-14.04.sh]: + [x] Done

Install open-appid by adding -a during the execution.

File: src/Snorter_Ubuntu-14.04.sh [hunk 10]
```diff
@@ -82,7 +102,24 @@ function snort_install() {
 	echo -ne "\n\t${CYAN}[i] INFO:${NOCOLOR} /var/log/snort and /etc/snort created and configurated.\n\n"
 	sudo /usr/local/bin/snort -V
 	echo -ne "\n\t${GREEN}[+] INFO:${NOCOLOR} ${BOLD}SNORT${NOCOLOR} is successfully installed and configurated!"
-
+	
+	#Installing OpenAppID
+	echo -ne "\n\t${CYAN}[i] INFO:${NOCOLOR} Installing ${BOLD}$OpenAppID${NOCOLOR}.\n\n"
+	cd $HOME/snort_src/
+	wget --no-check-certificate -P $HOME/snort_src/ https://snort.org/downloads/openappid/5411 -O snort-openappid.tar.gz
```
Comment (joanbono) [inline src/Snorter_Ubuntu-14.04.sh]: Will do something like this:

~~~~bash
OPENAPPVERSION=$(curl -s -k https://snort.org/downloads | grep -i "snort-openappid.tar.gz" | grep -oP "\/downloads\/openappid\/\d+")
~~~~ 

The output today (20170721) is:

~~~~bash
jbono@MacBook [~]> echo $OPENAPPVERSION
/downloads/openappid/5759
~~~~

Comment (joanbono) [inline src/Snorter_Ubuntu-14.04.sh]: + [x] Done

Also created a new function `install_openappid`, which will be only executed if the `-a` flag is set.

File: src/Snorter_Ubuntu-14.04.sh (hunk 1 above)
Comment (joanbono) [inline src/Snorter_Ubuntu-14.04.sh]: + [x] Done

**Comment by joanbono** [comment]
Hei bro @rbshadow 

I'll pushed all the fixes to the `devel` branch, please, test the installation. Run as follows:

~~~~bash
./Snorter -i $INTERFACE -o $OINKCODE -a
~~~~

So, the `-a` flag will install the OpenAppID 😄 

Test it and let me know. 

I'm going to close this PR.

**Comment by rbshadow** [comment]
Hi @joanbono bro.
I will let you know after a test. Thank you for awesome work.

**Comment by rbshadow** [comment]
Option doesn't work for.
`sudo ./Snorter_Ubuntu-14.04.sh -i eth0 -a --enabled-open-appid`

screenshot_15

**Comment by rbshadow** [comment]
Also for
`sudo ./Snorter_Ubuntu-14.04.sh -i eth0 -a`
Nghttp doesn't install first. It must be installed first before snort. I am trying to fix it. I tried several times but failed. It works perfectly when functions put in the main function serially. But if I put with a statement like 
`if [ -z $OPENAPPID ] ; then
    nghttp_install
fi
`
doesn't work also in main function.

**Comment by joanbono** [comment]
I think I know whats going wrong. I'll try to fix it now. Stay tuned.

**Comment by joanbono** [comment]
Now the `Snorter` main function is:

~~~~bash
function snort_install() {

	echo -ne "\n\t${CYAN}[i] INFO:${NOCOLOR} Installing dependencies.\n\n"
	sudo apt-get install -y --force-yes build-essential libpcap-dev libpcre3-dev libdumbnet-dev bison flex zlib1g-dev git locate vim 
	
	#Downloading DAQ and SNORT
	cd $HOME && mkdir snort_src && cd snort_src
	echo -ne "\n\t${CYAN}[i] INFO:${NOCOLOR} Downloading ${BOLD}$DAQ${NOCOLOR}.\n\n"
	wget --no-check-certificate -P $HOME/snort_src https://snort.org/downloads/snort/$DAQ.tar.gz
	echo -ne "\n\t${CYAN}[i] INFO:${NOCOLOR} Downloading ${BOLD}$SNORT${NOCOLOR}.\n\n"
	wget --no-check-certificate -P $HOME/snort_src https://snort.org/downloads/snort/$SNORT.tar.gz
	
	if [ -z "${OPENAPPID}" ] ; then
		nghttp_install
	fi
	
	#Installing DAQ
	cd $HOME/snort_src/
	echo -ne "\n\t${CYAN}[i] INFO:${NOCOLOR} Installing ${BOLD}$DAQ${NOCOLOR}.\n\n"
	tar xvfz $DAQ.tar.gz
	mv $HOME/snort_src/daq-*/ $HOME/snort_src/daq                     
	cd $HOME/snort_src/daq
	./configure && make && sudo make install 
	echo -ne "\n\t${GREEN}[+] INFO:${NOCOLOR} ${BOLD}$DAQ${NOCOLOR} installed successfully.\n\n"
	
	#Installing SNORT
	cd $HOME/snort_src/
	echo -ne "\n\t${CYAN}[i] INFO:${NOCOLOR} Installing ${BOLD}$SNORT${NOCOLOR}.\n\n"
	tar xvfz $SNORT.tar.gz > /dev/null 2>&1
	rm -r *.tar.gz > /dev/null 2>&1
	mv snort-*/ snort           
	cd snort
	./configure --enable-sourcefire $OPENAPPID
	make 
	sudo make install
	echo -ne "\n\t${GREEN}[+] INFO:${NOCOLOR} ${BOLD}$SNORT${NOCOLOR} installed successfully.\n\n"
	cd ..
	
	sudo ldconfig
	sudo ln -s /usr/local/bin/snort /usr/sbin/snort

	#Adding SNORT user and group for running SNORT
	echo -ne "\n\t${CYAN}[i] INFO:${NOCOLOR} Adding user and group ${BOLD}SNORT${NOCOLOR}.\n\n"
	sudo groupadd snort
	sudo useradd snort -r -s /sbin/nologin -c SNORT_IDS -g snort
	sudo mkdir /etc/snort > /dev/null 2>&1
	sudo mkdir /etc/snort/rules > /dev/null 2>&1
	sudo mkdir /etc/snort/preproc_rules > /dev/null 2>&1
	sudo touch /etc/snort/rules/white_list.rules /etc/snort/rules/black_list.rules /etc/snort/rules/local.rules > /dev/null 2>&1
	sudo mkdir /var/log/snort > /dev/null 2>&1
	sudo mkdir /usr/local/lib/snort_dynamicrules > /dev/null 2>&1
	sudo chmod -R 5775 /etc/snort > /dev/null 2>&1
	sudo chmod -R 5775 /var/log/snort > /dev/null 2>&1
	sudo chmod -R 5775 /usr/local/lib/snort_dynamicrules > /dev/null 2>&1
	sudo chown -R snort:snort /etc/snort > /dev/null 2>&1
	sudo chown -R snort:snort /var/log/snort > /dev/null 2>&1
	sudo chown -R snort:snort /usr/local/lib/snort_dynamicrules > /dev/null 2>&1
	
	sudo cp ~/snort_src/snort/etc/*.conf* /etc/snort > /dev/null 2>&1
	sudo cp ~/snort_src/snort/etc/*.map /etc/snort > /dev/null 2>&1
	
	sudo sed -i 's/include \$RULE\_PATH/#include \$RULE\_PATH/' /etc/snort/snort.conf
	
	echo -ne "\n\t${CYAN}[i] INFO:${NOCOLOR} /var/log/snort and /etc/snort created and configurated.\n\n"
	sudo /usr/local/bin/snort -V
	echo -ne "\n\t${GREEN}[+] INFO:${NOCOLOR} ${BOLD}SNORT${NOCOLOR} is successfully installed and configurated!"

	if [ -z "${OPENAPPID}" ] ; then
		install_openappid
	fi
}
~~~~

So, dependecies and folders are installed before install `nghttp` and also `Snort`.

**Comment by rbshadow** [comment]
Bro @joanbono I already tried as like your mentioned main function before your changes. But same things happen. Again I tried with your code now. It also doesn't work.

**Comment by joanbono** [comment]
Got the error:

The main function now is:

~~~~bash 
function snort_install() {

	echo -ne "\n\t${CYAN}[i] INFO:${NOCOLOR} Installing dependencies.\n\n"
	sudo apt-get install -y --force-yes build-essential libpcap-dev libpcre3-dev libdumbnet-dev bison flex zlib1g-dev git locate vim 
	
	#Downloading DAQ and SNORT
	cd $HOME && mkdir snort_src && cd snort_src
	echo -ne "\n\t${CYAN}[i] INFO:${NOCOLOR} Downloading ${BOLD}$DAQ${NOCOLOR}.\n\n"
	wget --no-check-certificate -P $HOME/snort_src https://snort.org/downloads/snort/$DAQ.tar.gz
	echo -ne "\n\t${CYAN}[i] INFO:${NOCOLOR} Downloading ${BOLD}$SNORT${NOCOLOR}.\n\n"
	wget --no-check-certificate -P $HOME/snort_src https://snort.org/downloads/snort/$SNORT.tar.gz
	
	if [ ! -z "${OPENAPPID}" ] ; then
		nghttp_install
	fi
	
	#Installing DAQ
	cd $HOME/snort_src/
	echo -ne "\n\t${CYAN}[i] INFO:${NOCOLOR} Installing ${BOLD}$DAQ${NOCOLOR}.\n\n"
	tar xvfz $DAQ.tar.gz
	mv $HOME/snort_src/daq-*/ $HOME/snort_src/daq                     
	cd $HOME/snort_src/daq
	./configure && make && sudo make install 
	echo -ne "\n\t${GREEN}[+] INFO:${NOCOLOR} ${BOLD}$DAQ${NOCOLOR} installed successfully.\n\n"
	
	#Installing SNORT
	cd $HOME/snort_src/
	echo -ne "\n\t${CYAN}[i] INFO:${NOCOLOR} Installing ${BOLD}$SNORT${NOCOLOR}.\n\n"
	tar xvfz $SNORT.tar.gz > /dev/null 2>&1
	rm -r *.tar.gz > /dev/null 2>&1
	mv snort-*/ snort           
	cd snort
	./configure --enable-sourcefire $OPENAPPID
	make 
	sudo make install
	echo -ne "\n\t${GREEN}[+] INFO:${NOCOLOR} ${BOLD}$SNORT${NOCOLOR} installed successfully.\n\n"
	cd ..
	
	sudo ldconfig
	sudo ln -s /usr/local/bin/snort /usr/sbin/snort

	#Adding SNORT user and group for running SNORT
	echo -ne "\n\t${CYAN}[i] INFO:${NOCOLOR} Adding user and group ${BOLD}SNORT${NOCOLOR}.\n\n"
	sudo groupadd snort
	sudo useradd snort -r -s /sbin/nologin -c SNORT_IDS -g snort
	sudo mkdir /etc/snort > /dev/null 2>&1
	sudo mkdir /etc/snort/rules > /dev/null 2>&1
	sudo mkdir /etc/snort/preproc_rules > /dev/null 2>&1
	sudo touch /etc/snort/rules/white_list.rules /etc/snort/rules/black_list.rules /etc/snort/rules/local.rules > /dev/null 2>&1
	sudo mkdir /var/log/snort > /dev/null 2>&1
	sudo mkdir /usr/local/lib/snort_dynamicrules > /dev/null 2>&1
	sudo chmod -R 5775 /etc/snort > /dev/null 2>&1
	sudo chmod -R 5775 /var/log/snort > /dev/null 2>&1
	sudo chmod -R 5775 /usr/local/lib/snort_dynamicrules > /dev/null 2>&1
	sudo chown -R snort:snort /etc/snort > /dev/null 2>&1
	sudo chown -R snort:snort /var/log/snort > /dev/null 2>&1
	sudo chown -R snort:snort /usr/local/lib/snort_dynamicrules > /dev/null 2>&1
	
	sudo cp ~/snort_src/snort/etc/*.conf* /etc/snort > /dev/null 2>&1
	sudo cp ~/snort_src/snort/etc/*.map /etc/snort > /dev/null 2>&1
	
	sudo sed -i 's/include \$RULE\_PATH/#include \$RULE\_PATH/' /etc/snort/snort.conf
	
	echo -ne "\n\t${CYAN}[i] INFO:${NOCOLOR} /var/log/snort and /etc/snort created and configurated.\n\n"
	sudo /usr/local/bin/snort -V
	echo -ne "\n\t${GREEN}[+] INFO:${NOCOLOR} ${BOLD}SNORT${NOCOLOR} is successfully installed and configurated!"

	if [ ! -z "${OPENAPPID}" ] ; then
		install_openappid
	fi
}
~~~~

So now will install the OpenAppID if the flag `-a` is set:

~~~~bash
	if [ ! -z "${OPENAPPID}" ] ; then
		install_openappid
	fi
~~~~ 

Test it, please

**Comment by joanbono** [comment]
Closing this PR, will work all this in an issue.

**Pull request closed without merging**